RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemoverPhase2.cc

```diff
@@ -0,0 +1,358 @@
+#include "DataFormats/TrackerRecHit2D/interface/ClusterRemovalInfo.h"
+
+#include "TrackingTools/PatternTools/interface/Trajectory.h"
+#include "TrackingTools/PatternTools/interface/TrajTrackAssociation.h"
+#include "TrackingTools/TrajectoryState/interface/TrajectoryStateOnSurface.h"
+#include "DataFormats/TrajectorySeed/interface/TrajectorySeedCollection.h"
+
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/CommonDetUnit/interface/GeomDetUnit.h"
+#include "Geometry/CommonDetUnit/interface/GeomDetType.h"
+
+//
+// class decleration
+//
+
+class SeedClusterRemoverPhase2 : public edm::stream::EDProducer<> {
+
+  public:
+    SeedClusterRemoverPhase2(const edm::ParameterSet& iConfig) ;
+    ~SeedClusterRemoverPhase2() ;
+    void produce(edm::Event &iEvent, const edm::EventSetup &iSetup) override ;
+  private:
+    bool doOuterTracker_, doPixel_;
+    bool mergeOld_;
+    typedef edm::ContainerMask<edmNew::DetSetVector<SiPixelCluster> > PixelMaskContainer;
+    typedef edm::ContainerMask<edmNew::DetSetVector<Phase2TrackerCluster1D> > Phase2OTMaskContainer;
+    edm::EDGetTokenT<edmNew::DetSetVector<SiPixelCluster> > pixelClusters_;
+    edm::EDGetTokenT<edmNew::DetSetVector<Phase2TrackerCluster1D> > phase2OTClusters_;
+    edm::EDGetTokenT<reco::ClusterRemovalInfo> oldRemovalInfo_;
+    edm::EDGetTokenT<PixelMaskContainer> oldPxlMaskToken_;
+    edm::EDGetTokenT<Phase2OTMaskContainer> oldPh2OTMaskToken_;
+    edm::EDGetTokenT<TrajectorySeedCollection> trajectories_;
+
+    std::vector<uint8_t> pixels, OTs;                // avoid unneed alloc/dealloc of this
+    edm::ProductID pixelSourceProdID, outerTrackerSourceProdID; // ProdIDs refs must point to (for consistency tests)
+
+    inline void process(const TrackingRecHit *hit, float chi2, const TrackerGeometry* tg);
+
+    template<typename T> 
+    std::auto_ptr<edmNew::DetSetVector<T> >
+    cleanup(const edmNew::DetSetVector<T> &oldClusters, const std::vector<uint8_t> &isGood, 
+                reco::ClusterRemovalInfo::Indices &refs, const reco::ClusterRemovalInfo::Indices *oldRefs) ;
+
+    // Carries in full removal info about a given det from oldRefs
+    void mergeOld(reco::ClusterRemovalInfo::Indices &refs, const reco::ClusterRemovalInfo::Indices &oldRefs) ;
+
+    bool clusterWasteSolution_;
+    bool filterTracks_;
+    reco::TrackBase::TrackQuality trackQuality_;
+    std::vector<bool> collectedOuterTrackers_;
+    std::vector<bool> collectedPixels_;
+};
+
+
+using namespace std;
+using namespace edm;
+using namespace reco;
+
+SeedClusterRemoverPhase2::SeedClusterRemoverPhase2(const ParameterSet& iConfig):
+    doOuterTracker_(iConfig.existsAs<bool>("doOuterTracker") ? iConfig.getParameter<bool>("doOuterTracker") : true),
+    doPixel_(iConfig.existsAs<bool>("doPixel") ? iConfig.getParameter<bool>("doPixel") : true),
+    mergeOld_(iConfig.exists("oldClusterRemovalInfo")),
```

Comment:
@slava77 - it looks like this part is cloned from the regular SeedClusterRemover.cc. I personally am against the cloning of algorithms like this, but for now this is what the tracking group has elected to do. I don't think the phase2 effort should become responsible for old code... maybe @VinInn could look at these separately?

(Re: one of your other comments, there are a bunch of cfi's whose names start with capital letters...)